app/blog/posts/how-to-make-authentication-with-github-apps-for-side-projects.md

```diff
@@ -0,0 +1,325 @@
+
+### Creating the node.js server
+
+Ensure that you have Node.js installed and create a new folder on your computer.
+
+We are going to use Express.js as it allows use some convenience with routing and redirects.
+
+```
+$ npm init -y
+$ npm install express esm
+```
+
+and we can create a `server.js`
+
+```javascript
+import express from 'express'
+
+const app = express()
+app.get('/', (req, res) => {
+  res.send('Hello GitHub auth')
+})
+const PORT = process.env.PORT || 9000
+app.listen(PORT, () => console.log('Listening on localhost:' + PORT))
+```
+
+Open `package.json` and add a new script to start the server
+
+```javascript
+"scripts": {
+	"start": "node -r esm server.js"
+}
+```
+
+We are using esm to use new ESNext `import` , which also works nicely with TypeScript if we want to convert it.
+
+### Saving GitHub App credentials
+
+We need to take the client id and client secret and store it in a way we can use it in our application.
+
+![GitHub App Credentials](images/github-auth/github-app-credentials.png)
+
+(You don't need to generate a private key now)
+
+We could store it in our code but that is a bad practice as we don't want to store secrets in our source code and it also makes it harder to change.
+
+A good start will be storing it as environment variables. We can use a npm package called `dotenv` which will load a file called `.env` and populate the process.env. This is smart as we can put that file in our .gitignore.
+
+```bash
+GITHUB_CLIENT_ID=yourid
+GITHUB_CLIENT_SECRET=yoursecret
+```
+
+and we can install dotenv
+
+```bash
+$ npm i dotenv
+```
+
+and we can update our `npm start` script to require dotenv
+
+```diff
+ "scripts": {
+-  "start": "node -r esm server.js"
++  "start": "node -r esm -r dotenv/config server.js"
+ }
+```
+
+and to test it you can now use this in your code:
+
+```
+console.log(process.env.GITHUB_CLIENT_ID)
+console.log(process.env.GITHUB_CLIENT_SECRET)
+```
+
+### Redirect to GitHub
+
+The first thing we need to add is redirect to GitHub. We do that by adding a new route.
+
+```javascript
+app.get('/login/github', (req, res) => {
+  const redirect_uri = 'http://localhost:9000/login/github/callback'
+  res.redirect(
+    `https://github.com/login/oauth/authorize?client_id=${
+      process.env.GITHUB_CLIENT_ID
+    }&redirect_uri=${redirect_uri}`
+  )
+})
+```
+
+After that if the user agrees with the GitHub prompt, she will be redirected back to the `redirect_url`.
+
+### Redirect route
+
+GitHub sends a `code` along as a URL query parameter, this we can grab with express with `req.query.code`.
+
+```javascript
+app.get('/login/github/callback', (req, res) => {
+  const code = req.query.code
+  // next step here
+})
+```
+
+We need to take that code and exchange it for a access token from the GitHub API. We can use the javascript `fetch` . I recommend using that in Node.js as well as it make you remember the api better for next time you want to use it in the browser.
```

Comment:
I don't really understand this well... 
Why do we exchange the `code`? What could be a use case for it? 
I also don't understand what you're recommending...

Reply:
Thank you for saying that, I will extend it.

The reason is that if GitHub returned it in the URL, somebody could steal it inbetween us, but since the "code" can't be used without the client_secret it is worthless to an attacker.

I recommend using fetch in node.js which is not our only choice. One could use the built in `http` module to do a http request or `axios`. The downside using something else than fetch is that you have to learn to different module apis.